### PR TITLE
Upgrade microsoft client and improve obj load test.

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR upgrades the microsoft client and adds some new checks to the object storage load test. The new checks for the object storage load test are some basic correctness tests for the object client functionality, which was used for essentially unit testing the changes to the microsoft client. Also, the chunking in the multipart upload functionality of the microsoft client was refactored into the ChunkWriteCloser abstraction (which I think may be useful in PutObjectAsync). The ChunkWriteCloser abstraction is not the most intuitive abstraction, but it is performant and simple. I would be interested in feedback on how similar performance/functionality could be accomplished.

### CODE REVIEW

File: src/client/pkg/grpcutil/stream.go [hunk 1]
```diff
@@ -49,6 +49,51 @@ func ChunkReader(r io.Reader, f func([]byte) error) (int, error) {
...
+	bufPool *BufPool
+	buf     []byte
+	f       func([]byte) error
+}
+
+// NewChunkWriteCloser creates a new ChunkWriteCloser.
+func NewChunkWriteCloser(bufPool *BufPool, f func(chunk []byte) error) *ChunkWriteCloser {
+	return &ChunkWriteCloser{
+		bufPool: bufPool,
+		buf:     bufPool.GetBuffer()[:0],
```
Comment (jdoliner) [inline src/client/pkg/grpcutil/stream.go]: Are we calling `PutBuffer` on this anywhere?

File: src/server/pkg/obj/microsoft_client.go [hunk 2]
```diff
@@ -34,72 +26,59 @@ var (
...
 	if err != nil {
 		return nil, err
 	}
-
-	return &microsoftClient{
-		blobClient: client.GetBlobService(),
-		container:  container,
-	}, nil
+	blobSvc := client.GetBlobService()
+	return &microsoftClient{container: (&blobSvc).GetContainerReference(container)}, nil
```
Comment (jdoliner) [inline src/server/pkg/obj/microsoft_client.go]: Wait so `client.GetBlobService` returns a non-pointer and you then need to take its address to call methods on it? That's pretty lame.

File: src/client/pkg/grpcutil/stream.go (hunk 1 above)
Comment (brycemcanally) [reply src/client/pkg/grpcutil/stream.go]: Yeah, it gets called in the callback (it is the responsibility of the caller, just like it is the responsibility of the caller to provide the buffer pool). This is not exactly the most intuitive pattern, but it allows the callback to kick off a goroutine(s) that uses the buffer and return it when it is done. The documentation for this abstraction explains this, and I take advantage of it in the microsoft client. The alternative, I think, is to implement a fix sized buffering abstraction (similar to `bufio.Writer`) then have the callback copy the data.

File: src/server/pkg/obj/microsoft_client.go (hunk 2 above)
Comment (brycemcanally) [reply src/server/pkg/obj/microsoft_client.go]: Yeah, I don't understand the reasoning for this.

File: src/server/pkg/obj/microsoft_client.go [hunk 3]
```diff
@@ -34,72 +26,59 @@ var (
...
+	blobRange := blobRange(offset, size)
+	if blobRange == nil {
+		return c.container.GetBlobReference(name).Get(nil)
 	}
+	return c.container.GetBlobReference(name).GetRange(&storage.GetBlobRangeOptions{Range: blobRange})
+}
 
-	if err != nil {
-		return nil, err
+func blobRange(offset, size uint64) *storage.BlobRange {
```
Comment (jdoliner) [inline src/server/pkg/obj/microsoft_client.go]: This is pretty minor, but want to have this return a `*storage.GetBlobRangeOptions`? I think that would make the above code a bit cleaner.

File: src/server/pkg/obj/microsoft_client.go [hunk 4]
```diff
@@ -135,103 +114,69 @@ func (c *microsoftClient) IsIgnorable(err error) bool {
...
 		return err
 	}
-	blocks := make([]storage.Block, w.nBlocks)
+	// Finalize the blocks.
+	blocks := make([]storage.Block, w.numBlocks)
 	for i := range blocks {
 		blocks[i] = storage.Block{ID: blockID(i), Status: storage.BlockStatusUncommitted}
 	}
-	return w.blobClient.PutBlockList(w.container, w.blob, blocks)
+	return w.blob.PutBlockList(blocks, nil)
```
Comment (jdoliner) [inline src/server/pkg/obj/microsoft_client.go]: I was hoping that the new azure-sdk would have support for contexts, but it looks like that's not the case. I just want to confirm to make sure I'm reading this right and that is indeed the case. It seems pretty weird they wouldn't have adopted context yet.

Comment (brycemcanally) [reply src/server/pkg/obj/microsoft_client.go]: Yeah, no support for contexts.

**Pull request merged**